File: src/main/java/hwr/oop/examples/polymorphism/parametric/Train.java
package hwr.oop.examples.polymorphism.parametric;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Objects;

class Train implements Comparable<Train> {
    private final Collection<Wagon> wagons;
    private final String id;

    public Train() {
        this("some train");
    }

    public Train(String id) {
        this.wagons = new ArrayList<>();
        this.id = id;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        Train train = (Train) o;
        return Objects.equals(wagons, train.wagons) &&
                Objects.equals(id, train.id);
    }

    @Override
    public int hashCode() {
        return Objects.hash(wagons, id);
    }

    @Override
    public int compareTo(Train o) {
        return this.id.compareTo(o.id);
    }
}

File: examples-kotlin/lecture-06/src/main/kotlin/hwr/oop/examples/polymorphism/adhoc/Train.kt
package hwr.oop.examples.polymorphism.adhoc

class Train(
  private val wagons: List<Wagon>,
) {
  constructor(vararg wagons: Wagon) : this(
    wagons.toList()
  )

//  constructor(wagons: Array<Wagon>): this(
//    wagons.toList()
//  )

  fun numberOfWagons(): Int {
    return wagons.size
  }
}
